File: python/purehtml/transformers/types/AttributeTransformer.py
from abc import ABC
from typing import List

from purehtml.transformers.Transformer import Transformer


class AttributeTransformer(Transformer, ABC):
    def __init__(self, args: List[str] = None):
        """
        Initialize the transformer with optional arguments.
        :param args: A list of attribute names to retrieve (can be empty for all attributes).
        """
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        """
        Set the list of attributes to retrieve.
        :param args: A list of attribute names.
        """
        self.args = args


    def get_name(self) -> str:
        """
        Return the name of the transformer.
        :return: The name of the transformer.
        """
        return "attr"

    def transform(self, params) -> str:
        """
        Retrieve the attributes of the HTML element.
        :param params: The parameters that contain the element to extract attributes from.
        :return: A dictionary of attributes.
        """
        element = params.get_element()
        attributes = {}

        if len(self.args) == 1:
            # If only one argument is provided, return the value of that attribute
            attribute_name = self.args[0].replace("(", "").replace(")", "")
            return element.get(attribute_name)

        elif not self.args:
            # If no arguments are provided, return all attributes
            for key, value in element.attrs.items():
                attributes[key] = value
        else:
            # If multiple arguments are provided, return the specified attributes
            for arg in self.args:
                attributes[arg] = element.get(arg)

        return str(attributes)

File: java/src/main/java/org/purescraps/purehtml/transformers/types/AttributeTransformer.java
package org.purescraps.purehtml.transformers.types;

import org.jsoup.nodes.Attribute;
import org.purescraps.purehtml.backend.PureHTMLNode;
import org.purescraps.purehtml.transformers.TransformParams;
import org.purescraps.purehtml.transformers.Transformer;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class AttributeTransformer extends Transformer {

    private List<String> args;

    // Constructor that accepts attribute names as arguments

    public AttributeTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    // Static method to return the name of the transformer
    public String getName() {
        return "attr";
    }

    // The transform method that retrieves the HTML element's attributes
    @Override
    public Object transform(TransformParams params) {
        /**
         * Retrieve the attributes of the HTML element.
         * @param params: The parameters that contain the element to extract attributes from.
         * @return: A map of attributes if multiple are requested, or a single attribute value.
         */
        PureHTMLNode element = params.getElement();
        Map<String, String> attributes = new HashMap<>();

        if (args.size() == 1) {
            // If only one argument is provided, return the value of that attribute
            String attributeName = args.getFirst().replace("(", "").replace(")", "");
            return element.attr(attributeName);
        } else if (args.isEmpty()) {
            // If no arguments are provided, return all attributes
            return element.attr().asList().stream()
                    .collect(Collectors.toMap(Attribute::getKey, Attribute::getValue));
        } else {
            // If multiple arguments are provided, return the specified attributes
            for (String arg : args) {
                attributes.put(arg, element.attr().get(arg)); // Assuming `get` fetches the attribute by name
            }
        }
        System.out.println(attributes);
        return attributes;
    }
}

